### PR TITLE
store/tikv: mock pd server and rpc server

### PR DESCRIPTION
**Pull request by disksing**
@shenli @morefreeze @coocood

### CODE REVIEW

File: store/tikv/mock-tikv/pd.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/pd/pd-client"
+)
+
+// Use global variables to prevent pdClients creating duplicate timestamps.
```
Comment (coocood) [inline store/tikv/mock-tikv/pd.go]: prevent pdClients **from** creating

File: store/tikv/mock-tikv/pd.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	tsMu       sync.Mutex
+	physicalTS int64
+	logicalTS  int64
+)
+
+type pdClient struct {
+	cluster *Cluster
+}
+
+// NewPDClient creates a mock pd.Client that uses lock timestamp and meta data
```
Comment (coocood) [inline store/tikv/mock-tikv/pd.go]: Do you mean **local** timestamp?

File: store/tikv/mock-tikv/rpc.go [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+func extractKeyErrors(errs []error) []*kvrpcpb.KeyError {
+	var errors []*kvrpcpb.KeyError
+	for _, err := range errs {
+		errors = append(errors, extractKeyError(err))
+	}
+	return errors
+}
+
+func extractPairs(pairs []Pair) []*kvrpcpb.KvPair {
```
Comment (coocood) [inline store/tikv/mock-tikv/rpc.go]: The name `extractPairs` doesn't properly decribe the work it does, it's more like convert pairs to `*kvrpcpb.KvPair`.

File: store/tikv/mock-tikv/pd.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		cluster: cluster,
+	}
+}
+
+func (c *pdClient) GetTS() (int64, int64, error) {
+	tsMu.Lock()
+	defer tsMu.Unlock()
+
+	ts := time.Now().UnixNano() / int64(time.Millisecond)
+	if physicalTS == ts {
```
Comment (shenli) [inline store/tikv/mock-tikv/pd.go]: physicalTS >= ts will be safer

File: store/tikv/mock-tikv/rpc.go [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+	return nil
+}
+
+func (h *rpcHandler) keyInRegion(key []byte) bool {
+	return regionContains(h.startKey, h.endKey, key)
+}
+
+func (h *rpcHandler) onGet(req *kvrpcpb.CmdGetRequest) *kvrpcpb.CmdGetResponse {
+	if !h.keyInRegion(req.Key) {
+		panic("onGet: key not in region")
```
Comment (shenli) [inline store/tikv/mock-tikv/rpc.go]: Why not return error in response?

File: store/tikv/mock-tikv/rpc.go [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+func (c *RPCClient) SendCopReq(req *coprocessor.Request) (*coprocessor.Response, error) {
+	return nil, errors.New("not implemented")
+}
+
+// Close closes the client.
+func (c *RPCClient) Close() error {
+	return nil
+}
+
+// NewRPCClient creates an RPCClient.
```
Comment (shenli) [inline store/tikv/mock-tikv/rpc.go]: an -> a

Comment (coocood) [inline store/tikv/mock-tikv/rpc.go]: @shenli 'R' starts with a vowel sound.

File: store/tikv/mock-tikv/rpc.go (hunk 3 above)
Comment (disksing) [inline store/tikv/mock-tikv/rpc.go]: @coocood The function names are ported from TiKV. But I agree it's not suitable here. How about `convertToKeyError` ?

File: store/tikv/mock-tikv/rpc.go (hunk 5 above)
Comment (disksing) [inline store/tikv/mock-tikv/rpc.go]: @shenli If the request passed the Region epoch check, but contains some keys not in the Region, there must be a bug in client's key split logic. I think panic could help us finding bug out as soon as possible.

Comment (morefreeze) [inline store/tikv/mock-tikv/rpc.go]: I think there must be something wrong in store/tikv, so fatal it.

**Comment by shenli** [comment]
LGTM

**Comment by coocood** [comment]
LGTM

**Pull request merged**